### PR TITLE
Added the method for adding Custom Script Extension to VMs in azure

### PR DESCRIPTION
**Pull request by aydey**

### CODE REVIEW

File: Tasks/DeployAzureResourceGroup/Environment.ts [hunk 1]
```diff
@@ -192,7 +191,7 @@ export class RegisterEnvironment {
             var fqdns = {}
             for (var i = 0; i < publicAddresses.length; i++) {
                 var publicAddress = publicAddresses[i];
-                var publicAddressId = publicAddress["id"];
+                var publicAddressId = publicAddress["id"]; //Didn't convert to lowercase
```
Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: It's an identifier used globally, conversion to lowercase is not necessary I guess

File: Tasks/DeployAzureResourceGroup/Environment.ts [hunk 2]
```diff
@@ -204,5 +203,124 @@ export class RegisterEnvironment {
...
+
+    private Add_WinRMHttpsNetworkSecurityRuleConfig(vmName: string){
+    }
+
+    public Add_AzureVMCustomScriptExtension(vmId: string, vmName: string, dnsName: string, location: string){
+        var _extensionName: string ="CustomScriptExtension"; 
+
+        console.log("Adding custom script extension for virtual machine %s",vmName);
+        console.log("VM Location: %s", location);
+        console.log("VM DNS: %s", dnsName);
```
Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: Localize the strings that you log

File: Tasks/DeployAzureResourceGroup/Environment.ts [hunk 3]
```diff
@@ -204,5 +203,124 @@ export class RegisterEnvironment {
...
+
+        try{
+            /*
+            Steps:
+                1. Check if CustomScriptExtension exists.
+                2. If not install it
+                3.Add the inbound rule to allow traffic on winrmHttpsPort
+            */
+            var computeClient = new computeManagementClient(this.credentials, this.subscriptionId);
+            var networkClient = new networkManagementClient(this.credentials, this.subscriptionId);
```
Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: You're not using this, are you?

File: Tasks/DeployAzureResourceGroup/Environment.ts [hunk 4]
```diff
@@ -204,5 +203,124 @@ export class RegisterEnvironment {
...
+            else{
+                this.Add_WinRMHttpsNetworkSecurityRuleConfig(vmName);
+            }
+        });
+    }
+
+    private _addExtensionVM(vmName: string, computeClient, dnsName: string, extensionName: string, location: string){
+        var _configWinRMScriptFile: string ="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/201-vm-winrm-windows/ConfigureWinRM.ps1";
+        var _makeCertFile: string ="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/201-vm-winrm-windows/makecert.exe";
+        var _winrmConfFile: string ="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/201-vm-winrm-windows/winrmconf.cmd";
```
Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: Are you sure about using raw.github links? 
What happens when there are any breaking changes?
Perhaps, you could stick to one commit ID

File: Tasks/DeployAzureResourceGroup/Environment.ts [hunk 5]
```diff
@@ -204,5 +203,124 @@ export class RegisterEnvironment {
...
+            this.Add_WinRMHttpsNetworkSecurityRuleConfig(vmName);
+        });
+    }
+
+    private _removeExtensionFromVM(extensionName, vmName, computeClient){
+        console.log("Removing the extension %s from vm %s", extensionName, vmName);
+        //delete the extension
+        computeClient.virtualMachineExtensions.deleteMethod(this.resourceGroupName, vmName, extensionName, (error, result, request, response)=>{
+            if(error){
+                console.log("Failed to delete the extension %s on the vm %s, with error Message: %s", extensionName, vmName, util.inspect(error, {depth: null}));
```
Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: Localize error logs

Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: You didn't set task results anywhere, you'll have to verify that as well.

File: Tasks/DeployAzureResourceGroup/Environment.ts (hunk 4 above)
Comment (aydey) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: The original task had it in this form... Let me know if I should change these

File: Tasks/DeployAzureResourceGroup/Environment.ts (hunk 1 above)
Comment (aydey) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: The original task used the id after converting it to lowercase, is the reason I put this comment

File: Tasks/DeployAzureResourceGroup/Environment.ts [hunk 6]
```diff
@@ -204,5 +205,207 @@ export class RegisterEnvironment {
...
+                networkClient.networkSecurityGroups.list(this.resourceGroupName, (error, result, request, response)=>{
+                    if(error){
+                        console.log("Error in getting the list of network Security Groups for the resource-group %s", this.resourceGroupName);
+                        throw "FetchingOfNetworkSecurityGroupFailed";
+                    }
+                    if(result.length > 0){
+                        console.log("Got network security group %s in resource group %s", securityGrpName, this.resourceGroupName);
+                        if(retryCnt>0){
+                            retryCnt = retryCnt - 1;
+                            this.AddInboundNetworkSecurityRule(retryCnt, securityGrpName, networkClient, ruleName, rulePriority, winrmHttpsPort);
```
Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: You can directly give this.AddInboundNetworkSecurityRule(retryCnt -1 , securityGrpName, networkClient, ruleName, rulePriority, winrmHttpsPort);

File: Tasks/DeployAzureResourceGroup/Environment.ts [hunk 7]
```diff
@@ -204,5 +205,207 @@ export class RegisterEnvironment {
...
+            }
+        }
+
+    private AddNetworkSecurityRuleConfig(securityGroups: [Object], ruleName: string, rulePriority: number, winrmHttpsPort: string){
+        for(var i =0; i < securityGroups.length; i++){
+            console.log("Adding Security rule for the network security group: %s", util.inspect(securityGroups[i], {depth: null}));
+            var securityGrp = securityGroups[i];
+            var securityGrpName = securityGrp["name"];
+
+            var maxtries = 3;
```
Comment (desattir) [inline Tasks/DeployAzureResourceGroup/Environment.ts]: You're not using this variable anywhere right?

**Pull request merged**